libtenzir/builtins/connectors/udp.cpp

```diff
@@ -0,0 +1,260 @@
+
+  sockaddr_in addr = {};
+  socket_type type = {};
+};
+
+/// Minimalistic RAII wrapper around a plain socket.
+struct socket {
+  socket() = default;
+
+  explicit socket(socket_type type) {
+    switch (type) {
+      case socket_type::invalid:
+        fd = -1;
+        break;
+      case socket_type::udp:
+        fd = ::socket(AF_INET, SOCK_DGRAM, 0);
+        break;
+      case socket_type::tcp:
+        fd = ::socket(AF_INET, SOCK_STREAM, 0);
+        break;
+    }
+  }
+
+  ~socket() {
+    if (*this) {
+      ::close(fd);
+    }
+  }
+
+  explicit operator bool() const {
+    return fd >= 0;
+  }
+
+  auto connect(socket_endpoint endpoint) {
+    auto* addr = endpoint.as_sockaddr();
+    constexpr socklen_t addr_len = sizeof(sockaddr_in);
+    return ::connect(fd, addr, addr_len);
+  }
+
+  auto bind(socket_endpoint endpoint) {
+    auto* addr = endpoint.as_sockaddr();
+    constexpr socklen_t addr_len = sizeof(sockaddr_in);
+    return ::bind(fd, addr, addr_len);
+  }
+
+  auto recv(std::span<std::byte> buffer, int flags = 0) {
+    return ::recv(fd, buffer.data(), buffer.size(), flags);
+  }
+
+  auto recvfrom(std::span<std::byte> buffer, socket_endpoint& endpoint,
+                int flags = 0) {
+    auto* addr = endpoint.as_sockaddr();
+    socklen_t addr_len = sizeof(sockaddr_in);
+    return ::recvfrom(fd, buffer.data(), buffer.size(), flags, addr, &addr_len);
+  }
+
+  int fd = -1;
+  socket_type type = socket_type::invalid;
+};
+
+auto udp_loader_impl(operator_control_plane& ctrl, loader_args args)
+  -> generator<chunk_ptr> {
+  auto sock = socket{socket_type::udp};
+  if (not sock) {
+    diagnostic::error("failed to create UDP socket")
+      .note("error: {}", detail::describe_errno())
+      .emit(ctrl.diagnostics());
+    co_return;
+  };
+  // A UDP packet contains its length as 16-bit field in the header, giving rise
+  // to packets sized up to 65,535 bytes (including the header). When we go
+  // over IPv4, we have a limit of 65,507 bytes (65,535 bytes − 8-byte UDP
+  // header − 20-byte IP header). At the moment we are not supporting IPv6
+  // jumbograms, which in theory get up to 2^32 - 1 bytes.
+  auto buffer = std::array<char, 65'536>{};
+  auto endpoint = socket_endpoint::parse(args.url);
+  if (not endpoint) {
+    diagnostic::error("invalid UDP endpoint")
+      .note("{}", endpoint.error())
+      .emit(ctrl.diagnostics());
+    co_return;
+  }
+  if (args.connect) {
+    TENZIR_DEBUG("connecting to {}", args.url);
+    if (sock.connect(*endpoint) < 0) {
+      diagnostic::error("failed to connect to socket")
+        .note("error: {}", detail::describe_errno())
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+  } else {
+    TENZIR_DEBUG("binding to {}", args.url);
+    if (sock.bind(*endpoint) < 0) {
+      diagnostic::error("failed to bind to socket")
+        .note("error: {}", detail::describe_errno())
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+  }
+  while (true) {
```

Comment:
You want to `co_yield {}` here once to signal that the operator started successfully.